winch/codegen/src/codegen/mod.rs

```diff
@@ -179,12 +266,24 @@ where
 
     /// Emit the usual function end instruction sequence.
     fn emit_end(&mut self) -> Result<()> {
-        self.context.pop_abi_results(&self.sig.result, self.masm);
         assert!(self.context.stack.len() == 0);
         self.masm.epilogue(self.context.frame.locals_size);
         Ok(())
     }
 
+    /// Returns the control stack frame at the given depth.
+    ///
+    /// # Panics
+    /// This function panics if the given depth cannot be associated
+    /// with a control stack frame.
+    pub fn control_at(frames: &mut [ControlStackFrame], depth: u32) -> &mut ControlStackFrame {
+        let index = (frames.len() - 1)
+            .checked_sub(depth as usize)
+            .unwrap_or_else(|| panic!("expected valid control stack frame at index: {}", depth));
```

Comment:
Does this need to handle `implicit` frames? Or is there an invariant here that only the outer body frame is implicit? If so, maybe we could remove the `implicit` flag (since it allows representing state that violates invariants) and instead handle the body-block case another way (e.g. at end/pop, recognize that it's the outermost block)?

Reply:
Yeah, really the only implicit frame that should exist should be the outermost. So this flag can be removed I believe and instead use the "outermost" frame as the invariant here.

Reply:
Updated!